servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/BeforeFinallyHttpOperator.java

```diff
@@ -170,15 +170,12 @@ public void onSuccess(@Nullable final StreamingHttpResponse response) {
                 // Invoking a terminal method multiple times is not allowed by the RS spec, so we assume we have been
                 // cancelled.
                 assert state == TERMINATED;
-                if (discardEventsAfterCancel) {
-                    return;
+                // Since we've been cancelled we need to drain the request body.
+                toSource(response.payloadBody()).subscribe(CancelImmediatelySubscriber.INSTANCE);
+                if (!discardEventsAfterCancel) {
+                    subscriber.onSuccess(response.transformMessageBody(payload ->
+                        Publisher.failed(new CancellationException("Received response post cancel."))));
                 }
-                subscriber.onSuccess(response.transformMessageBody(payload -> {
-                    // We have been cancelled. Subscribe and cancel the content so that we do not hold up the
-                    // connection and indicate that there is no one else that will subscribe.
-                    toSource(payload).subscribe(CancelImmediatelySubscriber.INSTANCE);
```

Comment:
Could we hold merging this before we discuss? There might be cases when we have to propagate as-is...

Reply:
Sure. And to be clear, the cleanup on the `discardEventsAfterCancel == false` pathway is has not changed in this PR. If we do want to change that, perhaps it's best done as a followup so as to not conflate the purpose of this PR.

Reply:
Thanks for clarifying. You are right, the `transformMessageBody` is not a lazy function, it's invoked right away and therefore it was also doing "subscribe-cancel" trick. I like that we moved it now, it makes it clear when it's invoked.

One more important thing here is that we never know what happened with the response message body before it reached this operator. It's possible that someone added another filter that aggregates it or drains. As a result, we risk to get a `DuplicateSubscribeException` here. I think we should play a safe game and try-catch this `subscribe-cancel`, because it's the best effort to drain.

Reply:
Actually, nvm. `DuplicateSubscribeException` is not thrown, it's delivered via `onError` and `CancelImmediatelySubscriber` will catch and log it.